### ISSUE TITLE
Extreme lag spikes after a short while of using the mod. Issue seems to happen after use of the Drawn Region mode.

### ISSUE THREAD

**Issue by Cephalaspis** [ChiselsAndBits/Chisels-and-Bits#583]
Before continuing if you are reporting an issue, please ensure that all mods involved are updated to the latest version available, if your issue with with intermod compatibility such mixing C&B bits with other blocks make sure that MCMP is also updated. If your issue is a crash, be sure to include the crash log with your report.

For any bug reports please provide the following details as well,
* MC Version: 1.16.4
* C&B Version: 0.2.9
* Do You have Optifine: Yes

**Comment by Cephalaspis** [comment]
UPDATE: After playing a little more, lag spikes became constant and not restricted to the use of the Drawn Region feature. I tried to copy a custom wall and paste it somewhere else, but the lag made it almost impossible.

**Comment by TacoSundae84** [comment]
The same thing is happening to me, but I don't have optifine. Disconnecting and reconnecting seems to fix the issue until after a few minutes of using this mod again. I was mostly using the negative design item, but it seemed to happen even when placing single bits.

**Comment by khan-22** [comment]
This is also happening to me. It looks like a memory leak. Checking F3 I can see memory usage slowly increase as I'm using c&b until it hits around 95%, then the game freezes for a moment, goes back to about 70%, and repeats. I increased my game's ram for now to 8 gigs which helped a bit.

**Comment by marchermans** [comment]
I have been trying to track this leak down.

I had it happen once to me during playing and testing. Which is why 0.2.9 took so long. Problem is that I could not reproduce it after that one time.

If anybody can get me a reproduction here that would help a lot.

I would need a way to reproduce it over and over and over again in just vanilla with c&b. 

Anybody have any ideas here? I am open for anything

**Comment by khan-22** [comment]
Forgot to mention that I'm also playing with the following mods on a 1.16.4/forge server:

- Optifine
- Cooking for blockheads
- Create 0.3
- Fairylights
- Fast leaf decay
- Just enough items
- Pam's Harvestcraft (all four modules)

The spikes started happening after I installed C&B and never before then, and it is the last one I installed. 

I'll have a quick go in singleplayer with only C&B enabled to see if the same thing happens.

**Comment by Cephalaspis** [comment]
I've recorded a video since I open the world to the moment the lag spikes appear (Around 5:50 approximately). You can see the amount of memory usage in the top right corner. I'm using Optifine though, so that might be an issue. 

[Video](https://youtu.be/MS4togGA9P4)

**Comment by khan-22** [comment]
Here's one instance of it happening in singleplayer with only C&B enabled (no optifine). I first tried invoking it by making the pit by chiseling away planes, but that wasn't very effective. I then tried to remove individual bits instead, which seemed to greatly increase memory usage (could just be that I reached a critical point just as I switched). I had to reach 99% a couple of times by chiseling for it to consistently spike from around 80% to 99%.

![image](https://user-images.githubusercontent.com/9250903/104821450-4131d900-583c-11eb-83c1-4cdc5ca174e7.png)

If I reload the world and do absolutely nothing it reaches a maximum of about 63%, I think. However, keep in mind that this is about 5 gigs for me, so if I had played on 2 gigs as I did before, then I guess I would have been faced with the lag immediately. Of course, if I start chiseling bits again I will eventually reach 99% as before.

**Comment by TacoSundae84** [comment]
I was trying to quickly get the memory leak up to a point where it caused lag (usually takes 10 to 20 minutes with normal use), and within a couple minutes of spamming the chisel everywhere, it started lagging really bad, memory was rapidly cycling between 90 and 70%, and then it crashed with this report.
https://pastebin.com/nAkE4mw4

**Comment by marchermans** [comment]
Thanks everybody.

I am going to do some digging tomorrow. Hopefully we can get to the bottom of this together.

**Comment by TacoSundae84** [comment]
One last observation that might be helpful - this happens even when not directly using this mod, but just building around it. I finished the parts i was using this mod for and put everything away earlier. But the memory problem is happening now from just looking at and placing blocks against chiseled blocks.

**Comment by MvatolokoS** [comment]
Happening on a small modpack I have aswell as previously stated I'm able to replicate it after a few minute of working on a piece (usually after breaking it to move it but sometimes after just a while of sculpting going to test a few things to see what info I can provide, I did have this problem the other day witha  small 1x1 block sculpture but i moved it then relogged and after being near it walking near it for a few days I didn't get any memory leak issues, it seems to be something with interacting with the piece with a chisel or with a bit. Again most reliable way for me to duplicate the problem is to try to break the item and move it. I start getting the leak the moment I break it. I am playing with optifine for 1.16.4 and the following mods

-Floocraft
-Tiny Mob Farm
-JEI
Fairy Lights
Forge 35.1.36
Pam's Harvest Craft 2
Extra Golems
Cooking for blockheads
Quark
AutoRegLib

I play with someone who is in the same area with me as I build something with the chisel and as I break it and they don't experience any memory leak what so ever.

**Comment by MvatolokoS** [comment]
Okay Did a bit of testing for about an hour and I found a way to replicate the problem/perhaps the cause of it. 

I got the tesselating crash at one point when placing one such that it interacts with another (they being two different complicated chisel structures)

I was able to replicate the memory leak by making two complicated chise and bit sculptures be on my screen at once. in singleplayer this does not cause a huge problem (thought ti still makes ram usage increase by about 15-20%) but in multiplayer this results with game breaking lag spikes due to a leak/spike in memory usage. All that is required is for the two items to be in view of the player IN ITEM FORM IN THE INVENTORY. For example if I only have one in my hotbar and I dont open my inventory only a 1-2% increase in ram might occur, Even if I break my other sculpture and pick it up WITHOUT OPENING MY INVENTORY TO VIEW IT everything remains fine, it isn't until I open my inventory then both items are on my screen, then you see my ram start to rise faster and faster until it starts peaking. 

Again it only seems to be gamebreaking in multiplayer but in singleplayer it still increases ram usage by about 20%. I have a video recording I can show of me trying various different things until I discovered what caused it.
